### PR TITLE
Fixes #23054 - Chain smart proxies to require the Foreman Service

### PR DESCRIPTION
**Pull request by ekohl**

### CODE REVIEW

**Comment by stbenjam** [comment]
I was working with a Satellite customer experiencing the behavior that during a foreman-installer run the foreman_smartproxy was registered before apache got started.  We tried this patch, and it didn't generate any autorequires in the debug log.  Weird that tests pass though, it seems to do the right thing here.

Does this require a specific Puppet release?

Editing the foreman_proxy module, and adding an explicit `requires => Service['httpd'],` to the register.pp file *did* fix the issue, so the underlying cause was the same.

**Comment by ekohl** [comment]
> I was working with a Satellite customer experiencing the behavior that during a foreman-installer run the foreman_smartproxy was registered before apache got started. We tried this patch, and it didn't generate any autorequires in the debug log. Weird that tests pass though, it seems to do the right thing here.

I have seen a similar problem in https://github.com/puppetlabs/puppetlabs-postgresql/pull/950. The tests pass but in acceptance tests I was able to produce errors that shouldn't happen. This could be a bug in Puppet. I'll need to dive deeper into this.

**Comment by stbenjam** [comment]
Hm, weird.  I cloned the BZ for the issue, it's here: http://projects.theforeman.org/issues/23054

**Comment by stbenjam** [comment]
Any chance you had a look into the autorequires problem? Saw another Satellite user hit this today.

**Comment by ekohl** [comment]
@stbenjam no, haven't had time to dive into this. Not sure if I'll have time soonish.

**Comment by ekohl** [comment]
@stbenjam I've pushed an updated version that explicitly requires httpd. Not too happy about it because it suggests this is incorrect (but there's no fix for that in this patch):
https://github.com/theforeman/puppet-foreman/blob/7728926d74e659fec4e3ba69ced5e7d844179dac/manifests/init.pp#L349-L351

Could you confirm this at least fixes the issue for the upgrading user?

File: lib/puppet/type/foreman_smartproxy.rb [hunk 1]
```diff
@@ -76,4 +76,13 @@ def refresh
...
 
+  autorequire(:class) do
+    ['Foreman::Service']
+  end
+
+  # puppetlabs-apache always defines the service httpd, even if the name is
+  # different.
+  autorequire(:service) do
+    ['httpd']
+  end
```
Comment (ehelms) [inline lib/puppet/type/foreman_smartproxy.rb]: I'm less familiar with autorequire -- does this continue to just work for proxies on servers without Apache?

**Comment by ekohl** [comment]
Alternative approach in https://github.com/theforeman/puppet-foreman/pull/700 which I believe is cleaner.

**Pull request closed without merging**